### ISSUE TITLE
[Bug]  zk's Secret will be set to empty when memory is not enougth

### ISSUE THREAD

**Issue by chentao11596** [strimzi/strimzi-kafka-operator#1663]
**Describe the bug**
When the next exception have been throw, zk's secret will be empty, and the next create or update for ca will be failed
<pre>
"2019-05-25 15:38:17 WARN  AbstractModel:368 - Error while generating certificates
java.io.IOException: Cannot allocate memory"

"2019-05-25 15:40:17 ERROR AbstractAssemblyOperator:176 - Reconciliation #233(timer) Kafka(test/chentao02-in-test): createOrUpdate failed
java.lang.NullPointerException: null"
</pre>

Follow the next codes:
<pre>
public Secret generateNodesSecret(ClusterCa clusterCa, Kafka kafka) {

        Map<String, String> data = new HashMap<>();

        log.debug("Generating certificates");
        Map<String, CertAndKey> certs;
        try {
            log.debug("Cluster communication certificates");
            certs = clusterCa.generateZkCerts(kafka);
            log.debug("End generating certificates");
            for (int i = 0; i < replicas; i++) {
                CertAndKey cert = certs.get(ZookeeperCluster.zookeeperPodName(cluster, i));
                data.put(ZookeeperCluster.zookeeperPodName(cluster, i) + ".key", cert.keyAsBase64String());
                data.put(ZookeeperCluster.zookeeperPodName(cluster, i) + ".crt", cert.certAsBase64String());
            }

        } catch (IOException e) {
            log.warn("Error while generating certificates", e);
        }

        return createSecret(ZookeeperCluster.nodesSecretName(cluster), data);
    }
</pre>
when clusterCa failed because of the java.io.IOException: Cannot allocate memory", the empty HashMap(data) will be used in the method "createSecret", so the empty happen

I think:
1) we no need to update this CA when exception occurs
2) for the log message "log.warn("Error while generating certificates", e);", we need to print kafka's info

**Expected behavior**

**Environment (please complete the following information):**
 - Strimzi version: 0.11.1
 - Installation method: YAML
 - Kubernetes cluster: Kubernetes 1.12.2
 - Infrastructure: KVM

**YAML files and logs**

Attach or copy paste the custom resources you used to deploy the Kafka cluster and the relevant YAMLs created by the Cluster Operator.
Attach or copy and paste also the relevant logs.

*To easily collect all YAMLs and logs, you can use our [report script](https://github.com/strimzi/strimzi-kafka-operator/blob/master/tools/report.sh) which will automatically collect all files and prepare a ZIP archive which can be easily attached to this issue.*

**Additional context**

**Comment by tombentley** [comment]
@chentao11596 thanks for the detailed report.

It's clear that exceptions thrown by `ClusterCa.generateZkCerts()` should not be caught, but instead propagated to the reconciliation loop. There's a similar problem with exceptions thrown by `ClusterCa.generateBrokerCerts()` being caught and logged. 

For the `java.io.IOException: Cannot allocate memory` specifically, it would be interesting to know where this is originating from and whether it's likely to be a transient problem. If it's likely to be thrown repeatedly once encountered it would probably be better for the CO process to exit and let the `Deployment` recreate the `Pod`.

**Comment by chentao11596** [comment]
It's happen because of the not enough memory which set by the pod. In fact, the ca can not re-created by himself even the pod recreate, so i think this should be a bug. I have to delete the secret (zk's) when i want to repair it.


<pre>
java.lang.NullPointerException: null
	at io.strimzi.operator.cluster.model.Ca.maybeCopyOrGenerateCerts(Ca.java:274) 

int replicasInSecret = secret == null || this.certRenewed() ? 0 : secret.getData().size() / 2;
</pre>

secret.getData() should be null, i think it will make the repair failed, thank you.

**Comment by tombentley** [comment]
Are you setting [resource requests](https://strimzi.io/docs/master/#assembly-resource-limits-and-requests-deployment-configuration-kafka) in your `Kafka` CR?

The most common way for lack of memory to affect the JVM is a `OutOfMemoryException`. The problem being via `IOException` suggests something like a NIO problem (i.e. mmap failed), or perhaps a problem creating a subprocess. Given this happens during cert generation, were we use `openssl`, the latter seems quite likely. I'd like to confirm that reasoning before concluding we should exit the CO process in this situation.

**Comment by chentao11596** [comment]
yes, i set it (256Mi). I have change it to avoid this appear again.

In my eyes, exit the process will be good.
